### PR TITLE
fix(share/discovery): decouple peer discovery from event processing

### PR DESCRIPTION
**Pull request by vgonkivs**
## Overview
Starting to listen to subscriptions async will help us to avoid any blocking in the case when we will not have the needed amount of FNs and will be blocked in `FindPeers`.
## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

**Review by renaynay — COMMENTED** [review]
let's test this on arabica bootstrappers

File: share/availability/discovery/discovery.go [hunk 1]
```diff
@@ -107,7 +108,7 @@ func (d *Discovery) EnsurePeers(ctx context.Context) {
 		return
 	}
 	// subscribe on Event Bus in order to catch disconnected peers and restart the discovery
-	sub, err := d.host.EventBus().Subscribe(&event.EvtPeerConnectednessChanged{})
+	sub, err := d.host.EventBus().Subscribe(&event.EvtPeerConnectednessChanged{}, eventbus.BufSize(512))
```
Comment (renaynay) [inline share/availability/discovery/discovery.go:111]: let's extract into a var in this file - Can we also document that the default buf size is 16 and why we increased it to 512

File: share/availability/discovery/discovery.go [hunk 2]
```diff
@@ -121,6 +122,32 @@ func (d *Discovery) EnsurePeers(ctx context.Context) {
 			log.Error(err)
 		}
 	}()
+
+	// starting to listen to subscriptions async will help us to avoid any blocking
+	// in the case when we will not have the needed amount of FNs and will be blocked in `FindPeers`.
+	go func() {
+		select {
+		case <-ctx.Done():
+			return
```
Comment (renaynay) [inline share/availability/discovery/discovery.go:131]: let's add some debug logs throughout this routine pls

File: share/availability/discovery/discovery.go (hunk 1 above)
Comment (Wondertan) [reply share/availability/discovery/discovery.go]: 512 is overkill btw, and 256 is more than enough, or even 128. There is literally nothing that could block event reading routine anymore

File: share/availability/discovery/discovery.go [hunk 3]
```diff
@@ -121,9 +129,40 @@ func (d *Discovery) EnsurePeers(ctx context.Context) {
 			log.Error(err)
 		}
 	}()
+
+	// starting to listen to subscriptions async will help us to avoid any blocking
+	// in the case when we will not have the needed amount of FNs and will be blocked in `FindPeers`.
+	go func() {
+		select {
+		case <-ctx.Done():
+			log.Debug("Context canceled. Finish listening for connectedness events.")
```
Comment (Wondertan) [inline share/availability/discovery/discovery.go:138]: Imo, it's pointless, but ok

File: share/availability/discovery/discovery.go [hunk 4]
```diff
@@ -121,9 +129,40 @@ func (d *Discovery) EnsurePeers(ctx context.Context) {
...
+			return
+		case e, ok := <-sub.Out():
+			if !ok {
+				log.Debug("Subscription for connectedness events is closed.")
```
Comment (Wondertan) [inline share/availability/discovery/discovery.go:142]: Same here.

**Review by Wondertan — APPROVED** [review]
Just two opinion farts, but LGTM

**Review by renaynay — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]

**Pull request merged**